### DISCUSSION TITLE
Add ability to build a cover letter

### DISCUSSION THREAD

**Discussion by achugr** [rendercv/rendercv#24] [Ideas]
A cover letter is often a good addition to a CV.
A cover letter is more personalized, but there is a template that can be reused from letter to letter. Plus, it would be great if the CV and the cover letter are of the same formatting and style.
What do you think about the idea of adding the ability to generate a cover letter based on a template? I would be happy to describe my proposal in more detail if you think this functionality fits the project.

**Comment by sinaatalay** [comment]
Hello, and thank you for the suggestion!

I don't think incorporating cover letters into the current codebase would be a good idea, as it would require an entirely different code. I think the current code won't help with cover letters, as they are purely related to CVs. Therefore, I believe creating a separate package for cover letters would make more sense. A single codebase that does two independent things may not be practical to maintain.

I might be wrong (maybe for users, it's better to have them in the same package), and I would like to hear your view on this. How would you implement this, and how would it look from the user side?

**Comment by achugr** [comment]
Hello!
Overall I see the process this way: you have two YAML configs - one for CV (the base one) and one for cover letter (the base one). When you want to apply for a position, you create a new branch from the main. You put there a custom cover letter and maybe adjust your CV according to position requirements. If you find some adjustments good and generic enough, you bring them to the main branch.
This does not require it to be the same code base, this could be a separate project rendercoverletter 🙈. I think a cover letter from the implementation perspective is much simpler than a CV and in terms of the code, this can be a much smaller data model and a simple template file. Having it in one repo makes it easy to reuse the code and keep default template styles in sync.
Other than that, this could be a separate project for a cover letter and an umbrella project that makes it easy to use both and makes sure styles (fonts etc) are in sync.

**Comment by achugr** [comment]
I'll prepare a draft PR in a couple of days and we could discuss there if it fits.

**Comment by sinaatalay** [comment]
I will keep this issue until we come up with a way of handling cover letters. I agree with you; it's important to have this.

**Comment by sinaatalay** [comment]
I am moving this issue to Discussions.

**Reply by achugr** [reply]
Does this mean we could try to find a place for this feature in the project?

**Reply by sinaatalay** [reply]
> Does this mean we could try to find a place for this feature in the project?

In my opinion, RenderCV should stay as only a CV generator tool for now. But it would be nice to hear what other people think about this.

**Reply by achugr** [reply]
How do you think, is there an easy way of bringing a cover letter in a different project, keeping in mind that it would be great to reuse template style (not necessary, but kinda cool) and put there some basic contact info (reuse data from CV config)?

**Reply by sinaatalay** [reply]
Another project focused on cover letters may be a good idea. RenderCV can be one of the dependencies of such a project. Also, TinyTeX of RenderCV can be used for rendering.

RenderCV’s reference is almost 100% written. I also use RenderCV as a package in the script I wrote for generating example entry images for the documentation. If you want to check it out, the script is in the `docs` folder.

The data models, templates, and functions of RenderCV can be leveraged into the features you mentioned very easily.

**Reply by achugr** [reply]
Could you please create a draft for it?
I'm not sure how to reuse this as a package, for example, `CurriculumVitae` is an attribute of `RenderCVDataModel` which is used in `LaTeXFile` for templating latex file, which makes rendering depend a lot on the data model `CurriculumVitae`.
I could probably extend `CurriculumVitae` with `CoverLetter` and use only fields I need in a template, so it becomes an "extended CV model that is displayed in a very reduced fashion in comparison to the base model", but this looks not too good conceptually. I am not too afraid of it, but maybe I miss something and it's all possible in a more elegant way?

**Comment by melissamcewen** [comment]
I created a cover letter with this, I'm not a Python dev so I wasn't sure how to extend as a dependency so I just hacked together a theme
https://github.com/melissamcewen/cv-studio/tree/main/engineeringCoverLetter
![John_Doe_CV_1](https://github.com/user-attachments/assets/3888a141-8413-4c2c-bdc8-212ed357a686)

Sample yml file: https://github.com/melissamcewen/cv-studio/blob/main/cl-test.yml

**Reply by achugr** [reply]
This is a smart solution, cool!

**Comment by sinaatalay** [comment]
Thank you for sharing this! That is a very nice idea. We should figure out a nice way of putting this in RenderCV.

**Reply by m3lixir** [reply]
It would be cool if each of the four given resume themes had their own matching cover letter theme 🙂

**Comment by DoctorD90** [comment]
Even though a couple of years have passed, just because I’ve noticed that this feature hasn’t been released yet (and if it has, I apologize; if you could point it out to me, I’d appreciate it!), I’d like to add my two cents on the subject, hoping to be of help <3

- `The cover letter is not part of the resume`: Honestly, I disagree; it often depends on the situation, but when a cover letter is written, it should be attached:
  - either as a separate file
  - or before the resume
  - or after the resume
- If you delegate the cover letter to third-party solutions, you run the risk of ending up with a different “design” (font, size, margins, color, etc.), and in my opinion, that would be quite annoying or might “convey” a lack of care for your “business card” to an HR representative (resume + cover letter).
- In almost all cases, a cover letter is a very simple piece of writing; in the past, when they were sent by post, they followed the standard letter format, with a header and greeting on the side, and sometimes a signature. Today, while it’s good practice that shows attention to detail, even a simple text is fine.

Personally, I've been using renderCV for a few days now, and the workaround I found is a custom section with `text entry` mode.

Based on what I’ve just explained, and with a view to keeping development as simple as possible, I would suggest:
- In the CV’s YAML file: a dedicated section, called, for example, `coverletter`, with attributes for the header (address, date, company, etc.), the closing greeting sentence (“Best regards,” “Kind regards,” etc.), the option to include an image (for a signature, in this case, for those who wish to do so), and the body of the text of course.
- In the design file settings for:
  - the TEMPLATE formula for the order of information in the `coverletter` sent-to-header (address, date, etc.)
  - the TEMPLATE formula for the order of information in the greeting (e.g., GREETING, NAME_IN_SNAKE)
  - sent-to-header position (right/left/centered)
  - position for greetings (right/left/centered)
  - whether to place it as:
    - A dedicated page, and in that case:
      - an option to decide whether to repeat the resume header (Name, address, phone number, contact information, etc.)
      - print it before or after the "cv" (sorry, I’m mainly envisioning the PDF as the output)
      - an option to export it to as separate file (it is very often requested as a separate file)
    - A standard section, and in that case:
      - where to place: where user enters the section in the YAML, first or last
      - an dedicated option to insert a page break before or after the section to allow the user to manage the layout (for example, if it’s 1 page and 2 lines, and you prefer to start the resume on a new page or immediately next, overriding others configuration for page split)

This way, it should just be a matter of adding a whole new section (but just simple text), and it shouldn't require any major changes. With the options to choose how to view or print it, users will have maximum flexibility.

What do you think?
